### PR TITLE
Proto serialization v25

### PR DESCRIPTION
**Pull request by dstrain115**
This PR creates a new class called CircuitSerializer that can serialize Circuits into protocol buffers more efficiently than existing mechanisms.

This class is similar to SerializableGateSet and op_(de)serializer, in that it serializes and deserializes cirq.Circuits into Program protocol buffers, but it is about 60% more space efficient and has significant time savings as well.

This serializer works by changing the Gate field and Args dict to a specific message for each allowed gate type.  It changes Qubits from a repeated Qubit message, to an index in the constant message.

This PR also adds a FloatArg message which is specific to Arguments where it is known the value will be a float (or a sympy expression).

### CODE REVIEW

File: cirq-google/cirq_google/circuit_serializer_test.py [hunk 1]
```diff
@@ -0,0 +1,673 @@
...
+
+X_PROTO = op_proto(
+    {
+        'xpowgate': {'exponent': {'float_value': 1.0}},
+        'qubit_constant_index': [0],
+    }
+)
+
+
+OPERATIONS = [
```
Comment (MichaelBroughton) [inline cirq-google/cirq_google/circuit_serializer_test.py:43]: Would it be possible to include symbol cases here and in the CircuitOperations tests ?

**Review by MichaelBroughton — CHANGES_REQUESTED** [review]
Looking good. A few minor nits. My biggest concern before merging is that the tests are totally comprehensive. Are we including all the cases with `sympy.Symbols` that we can in `circuit_serializer_test.py` ?

File: cirq-google/cirq_google/circuit_serializer.py [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+            )
+            if constants is not None:
+                msg.constants.extend(constants)
+            if arg_function_language is None:
+                arg_function_language = arg_func_langs._infer_function_language_from_circuit(
+                    msg.circuit
+                )
+        else:
+            raise NotImplementedError(f'Unrecognized program type: {type(program)}')
+        msg.language.arg_function_language = arg_function_language
```
Comment (maffoo) [inline cirq-google/cirq_google/circuit_serializer.py:107]: nit: I'd suggest putting this up by gate set at the top of the function.

File: cirq-google/cirq_google/circuit_serializer.py [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+        """Serialize a Circuit to cirq_google.api.v2.Program proto.
+
+        Args:
+            program: The Circuit to serialize.
+            msg: An optional proto object to populate with the serialization
+                results.
+            arg_function_language: The `arg_function_language` field from
+                `Program.Language`.
+            use_constants: Whether to use constants in serialization. This is
+                required to be True for serializing CircuitOperations.
```
Comment (maffoo) [inline cirq-google/cirq_google/circuit_serializer.py:84]: What would be the point of allowing this to be false? Could we remove this option and always use the constants table for v2.5?

File: cirq-google/cirq_google/circuit_serializer.py [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+            raw_constants: Optional[Dict[Any, int]] = {} if use_constants else None
+            self._serialize_circuit(
+                program,
+                msg.circuit,
+                arg_function_language=arg_function_language,
+                constants=constants,
+                raw_constants=raw_constants,
+            )
+            if constants is not None:
+                msg.constants.extend(constants)
```
Comment (maffoo) [inline cirq-google/cirq_google/circuit_serializer.py:100]: Could we instead pass `msg.constants` into `_serialize_circuit` so that the function can add constants to it directly? Perhaps this could be combined with `raw_constants` into a helper object that deals with the details of storing constants, since `constants` and `raw_constants` must both be present or not.

File: cirq-google/cirq_google/circuit_serializer.py [hunk 5]
```diff
@@ -0,0 +1,657 @@
...
+    ) -> Union[v2.program_pb2.Operation, v2.program_pb2.CircuitOperation]:
+        """Disambiguation for operation serialization."""
+        if msg is None:
+            if op.gate is not None:
+                return self.serialize_gate_op(op, msg, **kwargs)
+            if hasattr(op.untagged, 'circuit'):
+                return self.serialize_circuit_op(op, msg, **kwargs)
+            raise ValueError(f'Operation is of an unrecognized type: {op!r}')
+
+        if isinstance(msg, v2.program_pb2.Operation):
```
Comment (maffoo) [inline cirq-google/cirq_google/circuit_serializer.py:124]: It feels a bit awkward to have to check at runtime to match types between the proto message and the operation itself. I'd suggest having `_serialize_circuit` call `serialize_gate_op` or `serialize_circuit_op` directly since it already checks whether or not a given op is a `CircuitOperation`. I suppose this method could stay if you want it to be available to call directly (e.g. for tests), but I think it'd probably be fine to remove.

Also, I'd suggest moving `_serialize_circuit` up to be immediately after the `serialize` method; it's a bit hard to follow the flow when that internal method is located down after all the deserialization code.

File: cirq-google/cirq_google/circuit_serializer.py [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+            )
+            return msg
+        if isinstance(gate, cirq.WaitGate):
+            arg_func_langs.float_arg_to_proto(
+                gate.duration.total_nanos(),
+                out=msg.waitgate.duration_nanos,
+                arg_function_language=arg_function_language,
+            )
+            return msg
+        raise ValueError(f'Cannot serialize op {op!r} of type {type(gate)}')
```
Comment (maffoo) [inline cirq-google/cirq_google/circuit_serializer.py:285]: I would suggest changing this to a chain of `elif` with a final `else` that raises the exception, then having a single return at the end. Could also delay adding qubits and tags until after this check of gate types, since it doesn't make sense to do either of those if the gate type itself is not supported.

File: cirq-google/cirq_google/circuit_serializer.py [hunk 7]
```diff
@@ -0,0 +1,657 @@
...
+            arg_function_language: The `arg_function_language` field from
+                `Program.Language`.
+            constants: The list of previously-serialized Constant protos.
+            raw_constants: A map raw objects to their respective indices in
+                `constants`.
+
+        Returns:
+            The cirq.google.api.v2.CircuitOperation proto.
+        """
+        circuit = getattr(op.untagged, 'circuit', None)
```
Comment (maffoo) [inline cirq-google/cirq_google/circuit_serializer.py:311]: Should fail if `circuit` is None, so could remove the default here. But I would suggest instead doing an explicit type check:
```python
if not isinstance(op.untagged, cirq.CircuitOperation):
   raise ValueError(...)
circuit = op.untagged.circuit
```
This should also make the cast below unnecessary since we've already checked the type of `op.untagged`.

File: cirq-google/cirq_google/arg_func_langs.py [hunk 8]
```diff
@@ -198,54 +291,72 @@ def arg_from_proto(
...
+    *,
+    arg_function_language: str,
+    required_arg_name: Optional[str] = None,
+) -> Optional[ARG_LIKE]:
+    supported = SUPPORTED_FUNCTIONS_FOR_LANGUAGE.get(arg_function_language)
+    if supported is None:
+        raise ValueError(f'Unrecognized arg_function_language: {arg_function_language!r}')
 
-    if required_arg_name is not None:
+    if func.type not in cast(Set[str], supported):
```
[the remark this answers is not in the document]
Comment (dstrain115) [reply cirq-google/cirq_google/arg_func_langs.py]: Was from before.  Not really sure.  Removed.

File: cirq-google/cirq_google/circuit_serializer_test.py (hunk 1 above)
Comment (dstrain115) [reply cirq-google/cirq_google/circuit_serializer_test.py:43]: Done.  CircuitOperations with symbols still pending.

File: cirq-google/cirq_google/circuit_serializer.py [hunk 9]
```diff
@@ -0,0 +1,657 @@
...
+
+    This class is more performant than the previous `SerializableGateSet`
+    at the cost of some extendability.
+    """
+
+    def __init__(
+        self,
+        gate_set_name: str,
+    ):
+        """Construct the gate set.
```
[the remark this answers is not in the document]
Comment (dstrain115) [reply cirq-google/cirq_google/circuit_serializer.py]: Done.

File: cirq-google/cirq_google/circuit_serializer.py [hunk 10]
```diff
@@ -0,0 +1,657 @@
...
+        Args:
+            gate_set_name: The name used to identify the gate set.
+            serializers: The OpSerializers to use for serialization.
+                Multiple serializers for a given gate type are allowed and
+                will be checked for a given type in the order specified here.
+                This allows for a given gate type to be serialized into
+                different serialized form depending on the parameters of the
+                gate.
+            deserializers: The OpDeserializers to convert serialized
+                forms of gates or circuits into Operations.
```
[the remark this answers is not in the document]
Comment (dstrain115) [reply cirq-google/cirq_google/circuit_serializer.py]: Done.

File: cirq-google/cirq_google/circuit_serializer.py [hunk 11]
```diff
@@ -0,0 +1,657 @@
...
+                raw_constants=raw_constants,
+            )
+            if constants is not None:
+                msg.constants.extend(constants)
+            if arg_function_language is None:
+                arg_function_language = arg_func_langs._infer_function_language_from_circuit(
+                    msg.circuit
+                )
+        else:
+            raise NotImplementedError(f'Unrecognized program type: {type(program)}')
```
[the remark this answers is not in the document]
Comment (dstrain115) [reply cirq-google/cirq_google/circuit_serializer.py]: Done.

File: cirq-google/cirq_google/circuit_serializer.py (hunk 3 above)
Comment (dstrain115) [reply cirq-google/cirq_google/circuit_serializer.py]: Removed (for now).  My main concern was keeping the function signatures similar to SerializableGateSet, since the next change would be to add an interface that both of them follow, so that I can use both as a "gate_set".  I think I can get around this (or deprecate the use_constants in the other version?  In any case, I will deal with this in the follow-up PR when I unify the interface.

File: cirq-google/cirq_google/circuit_serializer.py (hunk 2 above)
Comment (dstrain115) [reply cirq-google/cirq_google/circuit_serializer.py]: Done.  Also got rid of the language inference.  There's no real need to iterate over the whole circuit again when we can just choose 'exp' and be done with it.

File: cirq-google/cirq_google/circuit_serializer.py (hunk 5 above)
Comment (dstrain115) [reply cirq-google/cirq_google/circuit_serializer.py]: Done.

File: cirq-google/cirq_google/circuit_serializer.py (hunk 6 above)
Comment (dstrain115) [reply cirq-google/cirq_google/circuit_serializer.py]: Done.

File: cirq-google/cirq_google/circuit_serializer.py (hunk 7 above)
Comment (dstrain115) [reply cirq-google/cirq_google/circuit_serializer.py]: Done.

File: cirq-google/cirq_google/circuit_serializer.py [hunk 12]
```diff
@@ -0,0 +1,657 @@
...
+        msg.language.arg_function_language = arg_function_language
+        return msg
+
+    def serialize_op(
+        self,
+        op: cirq.Operation,
+        msg: Union[None, v2.program_pb2.Operation, v2.program_pb2.CircuitOperation] = None,
+        **kwargs,
+    ) -> Union[v2.program_pb2.Operation, v2.program_pb2.CircuitOperation]:
+        """Disambiguation for operation serialization."""
```
[the remark this answers is not in the document]
Comment (dstrain115) [reply cirq-google/cirq_google/circuit_serializer.py]: Removed this function.

File: cirq-google/cirq_google/circuit_serializer.py [hunk 13]
```diff
@@ -0,0 +1,657 @@
...
+
+    def deserialize_op(
+        self,
+        operation_proto: Union[
+            v2.program_pb2.Operation,
+            v2.program_pb2.CircuitOperation,
+        ],
+        **kwargs,
+    ) -> cirq.Operation:
+        """Disambiguation for operation deserialization."""
```
[the remark this answers is not in the document]
Comment (dstrain115) [reply cirq-google/cirq_google/circuit_serializer.py:380]: REmoved this function.

File: cirq-google/cirq_google/circuit_serializer_test.py (hunk 1 above)
Comment (dstrain115) [reply cirq-google/cirq_google/circuit_serializer_test.py:55]: Modified circuit operation test to include symbols.

File: cirq-google/cirq_google/circuit_serializer.py (hunk 4 above)
Comment (dstrain115) [reply cirq-google/cirq_google/circuit_serializer.py]: Passed msg.constants into the file now.  Decided not to make a helper object given that raw_constants is the helper object now, and, with all the refactoring, everything that uses these is now private.

File: cirq-google/cirq_google/api/v2/program.proto [hunk 14]
```diff
@@ -106,17 +112,111 @@ message Language {
...
+}
+
+// Representation of cirq.XPowGate
+message XPowGate {
+  FloatArg exponent=1;
+}
+
+// Representation of cirq.YPowGate
+message YPowGate {
+  FloatArg exponent=1;
```
Comment (maffoo) [inline cirq-google/cirq_google/api/v2/program.proto:131]: nit: add whitespace around `=` for consistency, in several places in this file.

File: cirq-google/cirq_google/arg_func_langs.py [hunk 15]
```diff
@@ -35,11 +33,14 @@
...
     None: frozenset({'add', 'mul', 'pow'}),
 }
+MOST_PERMISSIVE_LANGUAGE = 'exp'
 
 SUPPORTED_SYMPY_OPS = (sympy.Symbol, sympy.Add, sympy.Mul, sympy.Pow)
 
 # Argument types for gates.
 ARG_LIKE = Union[int, float, List[bool], str, sympy.Symbol, sympy.Add, sympy.Mul]
+FLOAT_ARG_LIKE = Union[float, sympy.Symbol, sympy.Add, sympy.Mul]
+FLOAT_LIKE = (float, int, sympy.Integer, sympy.Float, sympy.Rational, sympy.NumberSymbol)
```
Comment (maffoo) [inline cirq-google/cirq_google/arg_func_langs.py:43]: nit: maybe name this differently to distinguish it from the `Union`s above, e.g. `FLOAT_LIKE_TYPES`. Might also be nice to add a comment to indicate why a tuple is needed here, instead of a union.

**Review by maffoo — APPROVED** [review]
A couple of nits, but otherwise LGTM!

File: cirq-google/cirq_google/arg_func_langs.py (hunk 15 above)
Comment (dstrain115) [reply cirq-google/cirq_google/arg_func_langs.py]: Done.

File: cirq-google/cirq_google/api/v2/program.proto (hunk 14 above)
Comment (dstrain115) [reply cirq-google/cirq_google/api/v2/program.proto]: Done.

**Comment by dstrain115** [comment]
Let me know if there are additional comments.  I will be submitting this week.

**Review by MichaelBroughton — APPROVED** [review]
